lib/mlxdevm.c

```diff
@@ -183,6 +183,55 @@ mlxdevm_sf_port_add(struct mlxdevm *dl, uint32_t pfnum, uint32_t sfnum)
 	return NULL;
 }
 
+static int cmd_port_dump_cb_to_list(const struct nlmsghdr *nlh, void *data)
+{
+	struct genlmsghdr *genl = mnl_nlmsg_get_payload(nlh);
+	struct nlattr *tb[MLXDEVM_ATTR_MAX + 1] = {};
+	struct mlxdevm_port_list_head *head = data;
+	struct mlxdevm_port_list *cur;
+	struct mlxdevm_port *port;
+	uint16_t flavour;
+
+	mnl_attr_parse(nlh, sizeof(*genl), attr_cb, tb);
+	if (!tb[MLXDEVM_ATTR_DEV_BUS_NAME] || !tb[MLXDEVM_ATTR_DEV_NAME] ||
+	    !tb[MLXDEVM_ATTR_PORT_INDEX] || !tb[MLXDEVM_ATTR_PORT_PCI_PF_NUMBER] ||
+	    !tb[MLXDEVM_ATTR_PORT_PCI_SF_NUMBER])
+		return MNL_CB_ERROR;
+
+	flavour = mnl_attr_get_u16(tb[MLXDEVM_ATTR_PORT_FLAVOUR]);
+	if (flavour != MLXDEVM_PORT_FLAVOUR_PCI_SF)
+		return MNL_CB_ERROR;
+
+	cur = calloc(1, sizeof(struct mlxdevm_port_list));
+	if (!cur)
+		return -ENOMEM;
+
+	port = &cur->port;
+	port->port_index = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_INDEX]);
+	port->ndev_ifindex = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_NETDEV_IFINDEX]);
+	port->pfnum = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_PCI_PF_NUMBER]);
+	port->sfnum = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_PCI_SF_NUMBER]);
+
+	cmd_port_fn_get(tb, port);
+
+	TAILQ_INSERT_TAIL(head, cur, entry);
+
+	return MNL_CB_OK;
+}
+
+int mlxdevm_sf_port_list_dump(struct mlxdevm *dl,
+			      struct mlxdevm_port_list_head *head)
```

Comment:
if this list not empty, return an error. otherwise you will add multiple entries twice.